tests/components/iaqualink/test_init.py

```diff
@@ -0,0 +1,360 @@
+        await hass.async_block_till_done()
+
+    assert config_entry.state == ConfigEntryState.SETUP_RETRY
+
+
+async def test_setup_all_good_no_recognized_devices(hass, config_entry):
+    """Test setup ending in no devices recognized."""
+    config_entry.add_to_hass(hass)
+
+    client = get_aqualink_client()
+    system = get_aqualink_system(client)
+    systems = {system.serial: system}
+    device = get_aqualink_device(system, AqualinkDevice)
+    devices = {device.name: device}
+
+    with patch(
+        "homeassistant.components.iaqualink.AqualinkClient.login", return_value=None
+    ), patch(
+        "homeassistant.components.iaqualink.AqualinkClient.get_systems",
+        return_value=systems,
+    ), patch(
+        "iaqualink.system.AqualinkSystem.get_devices", return_value=devices
+    ):
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    assert hass.data[DOMAIN][BINARY_SENSOR_DOMAIN] == []
+    assert hass.data[DOMAIN][CLIMATE_DOMAIN] == []
+    assert hass.data[DOMAIN][LIGHT_DOMAIN] == []
+    assert hass.data[DOMAIN][SENSOR_DOMAIN] == []
+    assert hass.data[DOMAIN][SWITCH_DOMAIN] == []
+
+    assert await hass.config_entries.async_unload(config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert config_entry.state == ConfigEntryState.NOT_LOADED
+    assert DOMAIN not in hass.data
+
+
+async def test_setup_all_good_all_device_types(hass, config_entry):
+    """Test setup ending in one device of each type recognized."""
+    config_entry.add_to_hass(hass)
+
+    client = get_aqualink_client()
+    system = get_aqualink_system(client)
+    systems = {system.serial: system}
+    devices = [
+        get_aqualink_device(system, AqualinkDevice),
+        get_aqualink_device(system, AqualinkAuxToggle),
+        get_aqualink_device(system, AqualinkBinarySensor),
+        get_aqualink_device(system, AqualinkLightToggle),
+        get_aqualink_device(system, AqualinkSensor),
+        get_aqualink_device(system, AqualinkThermostat),
+    ]
+    devices = {d.name: d for d in devices}
+
+    with patch(
+        "homeassistant.components.iaqualink.AqualinkClient.login", return_value=None
+    ), patch(
+        "homeassistant.components.iaqualink.AqualinkClient.get_systems",
+        return_value=systems,
+    ), patch(
+        "iaqualink.system.AqualinkSystem.get_devices", return_value=devices
+    ):
+        await hass.config_entries.async_setup(config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    assert config_entry.state == ConfigEntryState.LOADED
+
+    assert len(hass.data[DOMAIN][BINARY_SENSOR_DOMAIN]) == 1
+    assert len(hass.data[DOMAIN][CLIMATE_DOMAIN]) == 1
+    assert len(hass.data[DOMAIN][LIGHT_DOMAIN]) == 1
+    assert len(hass.data[DOMAIN][SENSOR_DOMAIN]) == 1
+    assert len(hass.data[DOMAIN][SWITCH_DOMAIN]) == 1
+
```

Comment:
https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

Please don't access `hass.data` in tests.  Assert that the entities are in the state machine instead